### PR TITLE
Bugfix: WinVaultKeyring.get_credential with non-existent username returns credential of other user (#698)

### PR DESCRIPTION
**Pull request by JamieBeverley**
Regarding #698

### CODE REVIEW

File: keyring/testing/backend.py [hunk 1]
```diff
@@ -181,3 +181,20 @@ def test_new_with_properties(self):
...
+        assert cred is None
+
+        password_1 = 'password1'
+        password_2 = 'password2'
+        self.set_password(service, 'user1', password_1)
+        self.set_password(service, 'user2', password_2)
+
+        assert keyring.get_credential(service, "user1").password == password_1
+        assert keyring.get_credential(service, "user2").password == password_2
+
```
Comment (JamieBeverley) [inline keyring/testing/backend.py]: Question/Note:
Considered adding this additional assertion too (see [previous commit)](https://github.com/jaraco/keyring/pull/699/commits/27d0e0ea3ea75ad7400e9d1d42bcbbbe088aaa01) but it was failing intermittently on linux.

Sometimes `keyring.get_credential(service, None).password ` would equate to `password_2` and sometimes to `password_1`.

Sticking a `time.sleep(1)` between the 2 `set_password` calls above helped it reliably pass so perhaps there's some async/timing thing happening with dbus (can possibly look into that deeper but a bit out of my depths w/ dbus)

```suggestion
        assert keyring.get_credential(service, "user2").password == password_2
         # Most recent entered
        assert keyring.get_credential(service, None).password == password_2
```

Comment (jaraco) [reply keyring/testing/backend.py]: Probably that additional assertion should be a separate test. Better to test one thing (the thing that matches the function's name). I'll use your comment to file a separate issue to track this concern (#700).

File: keyring/backends/Windows.py [hunk 2]
```diff
@@ -164,7 +164,8 @@ def get_credential(self, service, username):
             res = self._get_password(self._compound_name(username, service))
         # get any first password under the service name
```
Comment (jaraco) [inline keyring/backends/Windows.py]: This comment no longer matches the implementation.

File: keyring/backends/Windows.py [hunk 3]
```diff
@@ -164,7 +164,8 @@ def get_credential(self, service, username):
...
         if not res:
-            res = self._get_password(service)
-            if not res:
-                return None
+            cred = self._get_password(service)
+            res = cred if username is None or username == cred["UserName"] else None
+        if not res:
+            return None
```
Comment (jaraco) [inline keyring/backends/Windows.py]: I noticed these two lines were moved to the top level, but they are only relevant in this branch. Probably it makes sense not to change this behavior.

```suggestion
            if not res:
                return None
```

File: keyring/backends/Windows.py [hunk 4]
```diff
@@ -164,7 +164,8 @@ def get_credential(self, service, username):
             res = self._get_password(self._compound_name(username, service))
         # get any first password under the service name
         if not res:
-            res = self._get_password(service)
-            if not res:
-                return None
+            cred = self._get_password(service)
```
Comment (jaraco) [inline keyring/backends/Windows.py]: nitpick: the variable `cred` was used to represent the result from an API call (`_get_password`/`win32cred.CredRead`). Now `cred` is being used for the same purpose. I see why you did it, though. You needed a different name to represent the intermediate result to allow for `res` to represent the value when exiting the branch. Not sure there's anything to be done here.

File: keyring/backends/Windows.py (hunk 2 above)
Comment (jaraco) [reply keyring/backends/Windows.py]: Fixed in 2f3f4ec.

**Review by jaraco — APPROVED** [review]
This looks good. I'm going to do a bit of subsequent work to reorganize.

**Comment by jaraco** [comment]
I've always bemoaned that `get_credential` and `get_password` both had logic for resolving a credential. With 5e4b99e, that behavior is now unified and the tests still pass.

**Comment by JamieBeverley** [comment]
Thanks @jaraco - appreciate the comments + improvements!

**Pull request merged**